shared/js/content-scope/fingerprint.js

```diff
@@ -0,0 +1,111 @@
+    // eslint-disable-next-line new-cap
+    let hmac = new sjcl.misc.hmac(sjcl.codec.utf8String.toBits(sessionKey + domainKey), sjcl.hash.sha256)
+    return sjcl.codec.hex.fromBits(hmac.encrypt(inputData))
+}
+
+// linear feedback shift register to find a random approximation
+function nextRandom (v) {
+    return Math.abs((v >> 1) | (((v << 62) ^ (v << 61)) & (~(~0 << 63) << 62)))
+}
+
+let exemptionList = []
+
+function shouldExemptUrl (url) {
+    for (let regex of exemptionList) {
+        if (regex.test(url)) {
+            return true
+        }
+    }
+    return false
+}
+
+function initExemptionList (stringExemptionList) {
+    for (let stringExemption of stringExemptionList) {
+        exemptionList.push(new RegExp(stringExemption))
+    }
+}
+
+// Checks the stack trace if there are known libraries that are broken.
+function shouldExemptMethod () {
+    let errorLines = new Error().stack.split('\n')
+    let errorFiles = new Set()
+    // Should cater for Chrome and Firefox stacks, we only care about https? resources.
+    let lineTest = /(\()?(http[^)]+):[0-9]+:[0-9]+(\))?/
+    for (let line of errorLines) {
+        let res = line.match(lineTest)
+        if (res) {
+            let path = res[2]
+            // checked already
+            if (errorFiles.has(path)) {
+                continue
+            }
+            if (shouldExemptUrl(path)) {
+                console.log('Exempting script path:', path)
+                return true
+            }
+            errorFiles.add(res[2])
+        }
+    }
+    return false
+}
+
+// eslint-disable-next-line no-unused-vars
+function initCanvasProtection (args) {
+    let { sessionKey, stringExemptionList, site } = args
+    initExemptionList(stringExemptionList)
+    const domainKey = site.domain
+
+    // Using proxies here to swallow calls to toString etc
+    const getImageDataProxy = new Proxy(CanvasRenderingContext2D.prototype.getImageData, {
+        apply (target, thisArg, args) {
+            // The normal return value
+            const imageData = target.apply(thisArg, args)
+            if (shouldExemptMethod()) {
+                return imageData
+            }
+            const canvasKey = getCanvasKeySync(sessionKey, domainKey, imageData)
+            let pixel = canvasKey[0]
+            for (let i in canvasKey) {
+                let byte = canvasKey[i]
+                for (let j = 8; j >= 0; j--) {
+                    let pixelCanvasIndex = pixel % imageData.data.length
+
+                    imageData.data[pixelCanvasIndex] = imageData.data[pixelCanvasIndex] ^ (byte & 0x1)
+                    // find next pixel to perturb
+                    pixel = nextRandom(pixel)
+
+                    // Right shift as we use the least significant bit of it
+                    byte = byte >> 1
+                }
+            }
+            return imageData
+        }
+    })
+    CanvasRenderingContext2D.prototype.getImageData = getImageDataProxy
```

Comment:
What is the value of `CanvasRenderingContext2D.prototype.getImageData.toString()` after this?

Reply:
If we weren't using proxies, the content page can see the DDG code which is something we would like to prevent.

Reply:
So the proxy trick preserves the original `toString()` for this method? Nice.